### PR TITLE
feature(echo): support header filter and access phases.

### LINKED ISSUE #1225: feature: Support echo plugin
APISIX currently provides a simple example plugin, but it does not provide useful functionality. So we can provide a useful plugin to help users understand as fully as possible how to develop an APISIX plugin.

### PR DESCRIPTION
**Pull request by Akayeshmantha**
Fix #1225

### CODE REVIEW

**Comment by Akayeshmantha** [comment]
@agile6v

File: apisix/plugins/echo.lua [hunk 1]
```diff
@@ -70,4 +92,21 @@ function _M.body_filter(conf, ctx)
...
+    if conf.headers_arr then
+        local field_cnt = #conf.headers_arr
+        for i = 1, field_cnt, 2 do
+            ngx.header[conf.headers_arr[i]] = conf.headers_arr[i+1]
+        end
+    end
+end
+
+function _M.access(conf, ctx)
+    local auth_header = ngx.headers["Authorization"]
```
Comment (agile6v) [inline apisix/plugins/echo.lua:105]: It's best to specify the auth header.

File: apisix/plugins/echo.lua [hunk 2]
```diff
@@ -70,4 +92,21 @@ function _M.body_filter(conf, ctx)
...
+    if not auth_header then
+        return 401, { message = conf.unathorized_body }
```
Comment (agile6v) [inline apisix/plugins/echo.lua:107]: `unathorized_body ` need to be added in schema.

File: apisix/plugins/echo.lua [hunk 3]
```diff
@@ -70,4 +92,21 @@ function _M.body_filter(conf, ctx)
...
+    end
+    return 200, { message = conf.authorized_body }
```
Comment (agile6v) [inline apisix/plugins/echo.lua:109]: Can the value corresponding to the auth header be verified?

File: apisix/plugins/echo.lua [hunk 4]
```diff
@@ -52,6 +56,24 @@ local _M = {
 
 
 function _M.check_schema(conf)
+
+    if conf.headers then
```
Comment (agile6v) [inline apisix/plugins/echo.lua:60]: `headers` needs to be added in schema.

File: apisix/plugins/echo.lua (hunk 3 above)
Comment (Akayeshmantha) [reply apisix/plugins/echo.lua:109]: @agile6v  you mean like we do in the basic-auth plugin extract the values and check ?

File: apisix/plugins/echo.lua (hunk 4 above)
Comment (Akayeshmantha) [reply apisix/plugins/echo.lua:60]: yes will add them wanted to get an idea first weather the flow is correct.

Comment (Akayeshmantha) [reply apisix/plugins/echo.lua:77]: added

**Comment by Akayeshmantha** [comment]
@agile6v  pushed the changes 
@membphis

File: t/plugin/echo.t [hunk 5]
```diff
@@ -202,8 +212,32 @@ passed
...
+
+
+
+=== TEST 7: access with wrong auth header value throws 401
+--- request
+GET /hello
+--- more_headers
+Authorization: password
+--- error_code: 401
+--- response_body chomp
```
Comment (membphis) [inline t/plugin/echo.t:231]: need to check the response header

File: t/plugin/echo.t [hunk 6]
```diff
@@ -202,8 +212,32 @@ passed
...
+before the body modification unauthorized body
+--- wait: 0.2
+
+
+
+=== TEST 8: access without auth header value throws 401
+--- request
+GET /hello
+--- error_code: 401
```
Comment (membphis) [inline t/plugin/echo.t:240]: ditto

File: apisix/plugins/echo.lua [hunk 7]
```diff
@@ -31,6 +34,11 @@ local schema = {
         after_body = {
             description = "body after the modification of filter phase.",
             type = "string"
+        },
+        headers = {
+            description = "new headers for repsonse",
+            type = "object",
+            minProperties = 1,
```
Comment (membphis) [inline apisix/plugins/echo.lua:41]: can we only require the `headers`?

Comment (membphis) [reply apisix/plugins/echo.lua:41]: if the answer is YES, then we need to update the process: https://github.com/apache/incubator-apisix/pull/1708/files#diff-32537bd33cf9a44d9dd277b87d1ce5f9R82

File: apisix/plugins/echo.lua [hunk 8]
```diff
@@ -70,4 +94,22 @@ function _M.body_filter(conf, ctx)
     ngx.arg[2] = true
 end
 
+function _M.access(conf, ctx)
+    local auth_header = core.request.header(ctx, "Authorization")
+
+    if auth_header and auth_header == "userpass" then
+        return 200, "authorized body"
+    end
+    return 401, "unauthorized body"
```
Comment (membphis) [inline apisix/plugins/echo.lua:103]: I was confused about this process. @agile6v Do you think we should keep this?

File: t/plugin/echo.t (hunk 5 above)
Comment (Akayeshmantha) [reply t/plugin/echo.t:246]: done

File: apisix/plugins/echo.lua (hunk 7 above)
Comment (Akayeshmantha) [reply apisix/plugins/echo.lua:41]: I prefer to keep them since it is possible for someone to change the body of the response in the body filter part phase from what the user inputs. It is easy to understand that way.

File: apisix/plugins/echo.lua (hunk 8 above)
Comment (Akayeshmantha) [reply apisix/plugins/echo.lua]: So we talked about keeping this very simple. I thought maybe  to add the flow of basic auth plugin to extract username and PWD but we just want to convey the idea as per @agile6v. Like if someone need to access a protected resource in the access phase we need to have  some kind of an auth workflow 

WDYT @agile6v

Comment (agile6v) [reply apisix/plugins/echo.lua]: Hi @Akayeshmantha I mean we could specify the header and value, which is flexible and simple.

Comment (membphis) [reply apisix/plugins/echo.lua]: should we drop this code? https://github.com/apache/incubator-apisix/pull/1708/commits/dcbcc69a1c334761d01d3e85bfb3d60c35b90413#diff-32537bd33cf9a44d9dd277b87d1ce5f9R100-R104

Comment (Akayeshmantha) [reply apisix/plugins/echo.lua]: Hi @membphis  so I pushed the latest one

Comment (Akayeshmantha) [reply apisix/plugins/echo.lua]: of course with more tests.

File: apisix/plugins/echo.lua [hunk 9]
```diff
@@ -70,4 +102,23 @@ function _M.body_filter(conf, ctx)
     ngx.arg[2] = true
 end
 
+function _M.access(conf, ctx)
+    local auth_header = core.request.header(ctx, "Authorization")
+
+    if conf.auth_header == "Authorization" and conf.auth_value == "userpass" then
+            return 200, "authorized body"
+    end
```
Comment (membphis) [inline apisix/plugins/echo.lua:110]: `local auth_header` is it useless?

Comment (Akayeshmantha) [reply apisix/plugins/echo.lua:110]: @agile6v  wdyt ? 

I think at least we should showcase the workflow to pass auth header and then based on it give access in the access phase

Comment (agile6v) [reply apisix/plugins/echo.lua:110]: Yes. I have communicated with @Akayeshmantha on slack. And then review when the changes are completed?

Comment (membphis) [reply apisix/plugins/echo.lua:110]: ok, got it.

**Comment by membphis** [comment]
@Akayeshmantha I think you can add some doc things, then we can merge this PR soon

**Comment by Akayeshmantha** [comment]
@membphis  sure so I discussed with @agile6v  about giving proper documentation saying what this phase does and what we are showcasing from our examples. So, in the end, he mentioned let's finish up all the phases and start working on the documenting.

But already the plugin documentation in place.  

cc :- @agile6v

File: apisix/plugins/echo.lua [hunk 10]
```diff
@@ -70,4 +102,23 @@ function _M.body_filter(conf, ctx)
     ngx.arg[2] = true
 end
 
+function _M.access(conf, ctx)
+    local auth_header = core.request.header(ctx, "Authorization")
```
Comment (agile6v) [inline apisix/plugins/echo.lua:106]: This `core.request.header` function returns the value of the specified header. So it should be like this:
```
local value = core.request.header(ctx,  conf.auth_header)
if value ~= conf.auth_value then
    return 401, "unauthorized request."
end
```
In addition, the boundary conditions need to be considered.

Comment (Akayeshmantha) [reply apisix/plugins/echo.lua]: @agile6v

**Comment by Akayeshmantha** [comment]
@membphis  this error in travis not related to test cases. Ca we restart the job

**Comment by membphis** [comment]
I have restarted it.

**Comment by Akayeshmantha** [comment]
@membphis thank you

**Comment by Akayeshmantha** [comment]
@agile6v  @membphis  I think we can merge this right and will start on working the other phases

**Comment by membphis** [comment]
@Akayeshmantha merged, many thx
